### PR TITLE
[AutoDiff] Autodiff 6: Adstack regression tests

### PR DESCRIPTION
**Pull request by duburcqa**
# Adstack regression tests

> Test-only PR. Pins baseline reverse-mode AD behaviour through dynamic and static for-loops with the adstack extension. Complements Autodiff 1's unary-op test with coverage of the linear-kernel path, which has a different failure signature and different structural invariants.

## TL;DR

Three structural tests, each with an f32 and an f64 variant:

1. **`test_adstack_basic_gradient(_f64)`** — loop-carried state mutation. `v = v * 0.95 + 0.01` iterated `n_iter ∈ {1, 3, 10}` times. Gradient `dy/dx[i] = 0.95**n_iter`, independent of `x[i]`, so a broken push/load/pop pattern would still happen to produce the right answer (the backward chain `adj(v_prev) = 0.95 * adj(v_next)` only uses the compile-time constant 0.95). What this test pins is structural: the backward compiler must successfully reverse a dynamic `range(n_iter)`.

2. **`test_adstack_basic_gradient_negative`** — the negative counterpart. Decorated with `@test_utils.test(ad_stack_experimental_enabled=False)`; asserts `compute.grad()` raises `qd.QuadrantsCompilationError("Cannot use non static range in Backwards mode")` deterministically. Pins the exact compile-time rejection.

3. **`test_adstack_sum_linear(_f64)`** — 2×2×3 parametrize matrix: `(static vs dynamic loop) × (constant vs loop-varying coefficient) × (n_iter ∈ {1, 3, 10})`. Linear accumulation `y = sum_j v * coeff_j`; `v` is loop-invariant. Proves that enabling the adstack extension does not silently regress linear reverse-mode AD in any of the four shape combinations.

## Why

Autodiff 1 pins reverse-mode AD for *nonlinear* unary ops in dynamic loops — the shape where the per-iteration operand has to be replayed from the adstack. The class of bugs it catches: a unary op missing from `unary_collections` → operand falls back to a single spill slot → last-iteration value read for every backward step.

This PR pins the orthogonal shape: *linear* loop bodies. The failure modes here are different:

- The adstack is structurally required to reverse a dynamic range at all — without it, the backward compiler flat-out refuses to build the kernel. That's what `test_adstack_basic_gradient_negative` pins.
- Once the adstack enables the reversal, value-correctness of the spilled state depends on push/load/pop timing relative to the loop control flow. The linear-body tests do not exercise per-iteration value correctness of the stored `v` (the coefficients are compile-time constants), but they catch structural bugs that would stop the backward kernel from running at all — mis-sized slab, wrong iteration direction, off-by-one in the pop count on the reversed loop, and similar.

Both classes of test are needed: a unary-ops regression would slip past the linear tests, and a structural regression (e.g. an off-by-one that nonetheless happens to produce the right gradient for non-linear unary ops through compensation) would slip past the unary tests.

## The f64 design

Each test is a thin wrapper around a helper that takes `qd_dtype` and `rel_tol`:

```python
def _run_basic_gradient(qd_dtype, n_iter, rel_tol, shall_not_pass): ...
def _run_sum_linear(qd_dtype, use_static_loop, use_varying_coeff, n_iter, rel_tol): ...
```

The f32 and f64 tests decorate with different `@test_utils.test(...)` modes:

- f32: `@test_utils.test(require=qd.extension.adstack)`, `rel=1e-6` (tight for a strictly linear kernel; catches a new fp op introduced in the backward).
- f64: `@test_utils.test(require=[qd.extension.adstack, qd.extension.data64], default_fp=qd.f64)`, `rel=1e-14`.

`default_fp=qd.f64` at decorator time is load-bearing: Python float literals in the kernel body (`0.95`, `0.01`, etc.) get baked at whatever `default_fp` says, so `0.95` becomes an f32-quantized constant if the decorator uses the default `default_fp=qd.f32`, and the resulting gradient is not actually f64-precision. Splitting into two tests with different decorators is the only way to get truly tight f64 tolerance; the shared body lives in the helper.

## Coverage matrix

| Test                                       | Dynamic loop | Static loop | Loop-carried state | Check |
|--------------------------------------------|:-:|:-:|:-:|---|
| `test_adstack_basic_gradient`              | ✓ |   | ✓ | value match against `0.95**n_iter` |
| `test_adstack_basic_gradient_negative`     | ✓ |   | ✓ | compile-time rejection |
| `test_adstack_sum_linear[static, const]`   |   | ✓ |   | value match against `n_iter` |
| `test_adstack_sum_linear[static, varying]` |   | ✓ |   | value match against `sum(a+1 for a)` |
| `test_adstack_sum_linear[dyn, const]`      | ✓ |   |   | same |
| `test_adstack_sum_linear[dyn, varying]`    | ✓ |   |   | same |

## Stack

Autodiff 6 of 13. Based on #496 (alloca placement). Followed by #534 (header size).

### CODE REVIEW

**Comment by duburcqa** [comment]
Addressed in 5498454ad. The Codex review correctly identified that UnaryOpType::tan was missing from NonLinearOps::unary_collections, but fixing only the set membership does not actually fix the bug it describes: a test with `a = tan(a)` in a dynamic loop still returns a wrong gradient because MakeAdjoint's tan formula was `adj(stmt) * (1 + sqr(stmt))`, and BackupSSA spills the forward tan result (stmt) to a plain alloca, so the reversed loop reads the last-iteration value for every backward step. The same latent bug also affected tanh and exp. The fix is therefore two-part: add tan to the non-linear set so the operand's alloca is promoted to an AdStack, and rewrite tan/tanh/exp's reverse formula to recompute the op on the adstack-backed operand rather than reuse the forward stmt. Covered by the new parametrized test_adstack_unary_loop_carried, which uses torch.autograd as the reference.

**Comment by duburcqa** [comment]
@claude review

**Comment by duburcqa** [comment]
@claude review

File: quadrants/transforms/auto_diff.cpp [hunk 1]
```diff
@@ -1246,11 +1259,11 @@ class MakeAdjoint : public ADTransform {
...
-      // `tan`.
-      QD_NOT_IMPLEMENTED;
+      // d/dx tan(x) = 1 + tan(x)^2. Uses the already-computed tan value rather than a fresh
+      // 1/cos(x)^2, which is both cheaper and numerically identical. The derivative diverges
+      // at x = pi/2 + k*pi; NaN propagation matches how sqrt, log, asin, acos, rsqrt handle
+      // their own singular inputs.
+      accumulate(stmt->operand, mul(adjoint(stmt), add(constant(1, stmt->ret_type), sqr(stmt))));
     } else if (stmt->op_type == UnaryOpType::tanh) {
       accumulate(stmt->operand, mul(adjoint(stmt), sub(constant(1, stmt->ret_type), sqr(stmt))));
     } else if (stmt->op_type == UnaryOpType::asin) {
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply quadrants/transforms/auto_diff.cpp:1312]: Fixed by splitting the PR: `UnaryOpType::tan` is now added to `NonLinearOps::unary_collections` in #501, and the deep regression coverage lives in `test_adstack_unary_loop_carried` on #500 (whose parametrize list #501 extends with `qd.tan`). That test uses a redefinition pattern (`a = x[i] + qd.cast(j, qd.f32) * 0.1`) so there is no `LocalStore(a)` after `LocalLoad(a)` in the inner body, and the `cast + add` chain is outside `binary_collections` - promotion hinges solely on `unary_collections` membership, so dropping `tan` from the set makes the test fail.

File: quadrants/transforms/auto_diff.cpp [hunk 2]
```diff
@@ -41,9 +41,9 @@ class IndependentBlockMetaData {
 class NonLinearOps {
  public:
   inline static const std::set<TernaryOpType> ternary_collections{TernaryOpType::select};
-  inline static const std::set<UnaryOpType> unary_collections{UnaryOpType::abs,  UnaryOpType::sin,  UnaryOpType::cos,
-                                                              UnaryOpType::tanh, UnaryOpType::asin, UnaryOpType::acos,
-                                                              UnaryOpType::exp,  UnaryOpType::log,  UnaryOpType::sqrt};
+  inline static const std::set<UnaryOpType> unary_collections{
+      UnaryOpType::abs,  UnaryOpType::sin,  UnaryOpType::cos, UnaryOpType::tan, UnaryOpType::tanh,
+      UnaryOpType::asin, UnaryOpType::acos, UnaryOpType::exp, UnaryOpType::log, UnaryOpType::sqrt};
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply quadrants/transforms/auto_diff.cpp]: Fixed by splitting the PR: `UnaryOpType::rsqrt` is now added to `NonLinearOps::unary_collections` in #503, and the reviewer's concern about the false-confidence update (`a = a * 0.99 + 0.01` promoting via `mul`) is addressed by the redefinition pattern in `test_adstack_unary_loop_carried` (on #500): `a = x[i] + qd.cast(j, qd.f32) * 0.1` has no `LocalStore(a)` after `LocalLoad(a)` and uses only `cast + add` (both outside `binary_collections`), so promotion hinges solely on `unary_collections` membership. #503 extends that test's parametrize with `qd.rsqrt`; flipping `rsqrt` out of the set makes the test fail.

File: tests/python/test_adstack.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+def test_adstack_unary_loop_carried(qd_op, torch_op):
+    # The update on `a` is purely additive (no mul/div/pow) so the nonlinear unary op under test is the SOLE consumer
+    # that could trigger AdStackAllocaJudger to promote the loop-carried alloca. A multiplicative update like
+    # `a = a * 0.99 + ...` would promote the alloca through the mul, giving false confidence about AdStack coverage
+    # for unary ops.
+    import torch
+
+    x = qd.field(qd.f32)
+    y = qd.field(qd.f32)
+    qd.root.dense(qd.i, 1).place(x, x.grad)
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply tests/python/test_adstack.py]: Fixed: the test was rewritten on #500 to use a redefinition pattern (`a = x[i] + qd.cast(j, qd.f32) * 0.1`) instead of an additive update. There is no `LocalStore(a)` after `LocalLoad(a)` in the inner body, and the `cast + add` chain is outside `binary_collections`, so the `LocalStore-after-LocalLoad` promotion path the reviewer flagged can no longer trigger. Promotion genuinely hinges on `NonLinearOps::unary_collections` membership, and dropping any parametrized op from the set makes the test fail.

File: tests/python/test_adstack.py [hunk 4]
```diff
@@ -5,6 +5,100 @@
 from tests import test_utils
 
 
+@test_utils.test(require=qd.extension.adstack, ad_stack_experimental_enabled=True)
+def test_adstack_basic_gradient():
```
Comment (hughperkins) [inline tests/python/test_adstack.py]: would be good to add a docstring to each test saying what is special about the test, what kind of issue it might detect.

**Comment by hughperkins** [comment]
Opus description:

  ## Summary
  Adds four new adstack regression tests that exercise reverse-mode AD through dynamic
  (and static) loops with simple linear operations — no unary math ops involved. These
  complement the existing `test_adstack_unary_loop_carried` tests by covering the basic
  loop-carried accumulation and scaling patterns that the adstack must handle correctly.
  ### Tests added
  - **`test_adstack_basic_gradient`:** A loop-carried decay (`v = v * 0.95 + 0.01`) over
    10 iterations. Checks that `x.grad` equals `0.95^10`, verifying the chain rule through
    repeated linear updates.
  - **`test_adstack_sum_fixed_coeff`:** Sums `v = x[i]` three times inside a dynamic loop
    (`y += v`). Checks `x.grad == 3.0` — the simplest possible loop-carried gradient.
  - **`test_adstack_sum_constant_coeffs`:** Sums `v * (a+1)` for `a` in `range(3)`, where
    the coefficient varies per iteration but `v` is loop-invariant. Checks
    `x.grad == 1 + 2 + 3 = 6.0`.
  - **`test_adstack_sum_static_loop_correct`:** Same as `test_adstack_sum_fixed_coeff` but
    uses `qd.static(range(3))` instead of a dynamic range, verifying that static loop
    unrolling doesn't break the gradient.
  ## Strengths
  - **Covers the linear/affine case.** The existing tests all hinge on non-linear unary
    ops; these four tests verify that the adstack works correctly even when the loop body
    is purely linear, which is arguably the more common real-world pattern.
  - **Easy to reason about.** Every expected gradient is a simple closed-form expression
    (`0.95^10`, `3.0`, `6.0`), making failures immediately interpretable.
  - **Dynamic vs. static coverage.** Testing both `range(3)` and `qd.static(range(3))`
    guards against regressions where one path works and the other doesn't.
  - **Progressive complexity.** The four tests form a natural ladder: fixed read → fixed
    read with scaling → loop-varying coefficient → loop-carried state mutation.
  ## Weaknesses / things to consider
  - **No docstrings.** None of the four tests have comments explaining what they're
    testing or why. The existing tests in this file have detailed comments; these would
    benefit from at least a one-line description each.
  - **Single-element fields throughout.** All four tests use `dense(qd.i, 1)`. The
    existing `test_adstack_unary_loop_carried` was upgraded to use 4 elements — these
    could do the same for consistency and to catch per-element accumulation bugs.
  - **`ad_stack_experimental_enabled=True` but no `require=qd.extension.adstack`…**
    Actually, all four tests do have `require=qd.extension.adstack` — disregard. However,
    unlike the unary tests, none of these use `@pytest.mark.needs_torch` or
    cross-validate against PyTorch, which is fine since the expected values are trivially
    hand-computed.
  - **Fixed loop counts only.** Unlike the unary tests which parametrize over
    `n_iter=[1, 3, 10]`, these tests use a single hardcoded loop count each. Parametrizing
    would increase coverage cheaply.
  - **`test_adstack_basic_gradient` is the only test with loop-carried state mutation**
    (`v = v * 0.95 + 0.01`). The other three read `v` without modifying it in the loop,
    so they don't truly exercise adstack replay of evolving state — they could pass even
    without adstack promotion.

**Comment by hughperkins** [comment]
Please address Opus commets:
- No docstrings. None of the four tests have comments explaining what they're
testing or why. The existing tests in this file have detailed comments; these would
benefit from at least a one-line description each.
- Single-element fields throughout. All four tests use dense(qd.i, 1). The
existing test_adstack_unary_loop_carried was upgraded to use 4 elements — these
could do the same for consistency and to catch per-element accumulation bugs.
- ad_stack_experimental_enabled=True but no require=qd.extension.adstack…
Actually, all four tests do have require=qd.extension.adstack — disregard. However,
unlike the unary tests, none of these use @pytest.mark.needs_torch or
cross-validate against PyTorch, which is fine since the expected values are trivially
hand-computed.
- Fixed loop counts only. Unlike the unary tests which parametrize over
n_iter=[1, 3, 10], these tests use a single hardcoded loop count each. Parametrizing
would increase coverage cheaply.
- test_adstack_basic_gradient is the only test with loop-carried state mutation
(v = v * 0.95 + 0.01). The other three read v without modifying it in the loop,
so they don't truly exercise adstack replay of evolving state — they could pass even
without adstack promotion.

In addition, can we add a 'negative test' in each unit test please:
- disable adstack
    - check gradients are zero

File: tests/python/test_adstack.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (duburcqa) [reply tests/python/test_adstack.py]: Addressed: removed the redundant `ad_stack_experimental_enabled=True` from all four decorators on this PR (and cascaded the same cleanup through #495, #490, #493). `test_utils.test` auto-enables the option when `qd.extension.adstack` appears in `require` - see `tests/test_utils.py:250-251`.

File: tests/python/test_adstack.py [hunk 5]
```diff
@@ -140,3 +141,120 @@ def compute():
...
+    # separate tests (`test_adstack_sum_fixed_coeff`, `test_adstack_sum_constant_coeffs`,
+    # `test_adstack_sum_static_loop_correct`) with a single parametrized version so every branch of that truth
+    # table is covered at each trip count.
+    #
+    # Internal details: this test deliberately does not mutate `v` inside the loop, so the reverse pass does not
+    # require adstack replay of `v` to compute the right gradient - `v`'s per-iteration value is the same `x[i]`.
+    # The point of this test is therefore not to stress the adstack (that is `test_adstack_basic_gradient`'s job)
+    # but to prove that enabling the adstack extension does not silently regress linear reverse-mode AD for either
+    # unrolled or dynamic loop shapes. There is no negative counterpart because disabling the adstack is the
+    # already-correct path for these kernels - the expected gradient is the same whether the option is on or off.
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply tests/python/test_adstack.py]: Rewrote both docstrings to match the actual semantics: `test_adstack_basic_gradient_negative` now says it pins the compile-time `QuadrantsCompilationError`; `test_adstack_sum_linear` now qualifies the no-negative-counterpart claim as applying only to `use_static_loop=True` (with `use_static_loop=False` the compile error covered by the basic_gradient negative test is out of scope here).

File: tests/python/test_adstack.py [hunk 6]
```diff
@@ -3,6 +3,7 @@
 import pytest
 
 import quadrants as qd
+from quadrants.lang.exception import QuadrantsCompilationError
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply tests/python/test_adstack.py]: Dropped the `from quadrants.lang.exception import QuadrantsCompilationError` import and now reference `qd.QuadrantsCompilationError` directly in the `pytest.raises` call.

File: tests/python/test_adstack.py [hunk 7]
```diff
@@ -140,3 +140,120 @@ def compute():
...
+    # `0.95 ** n_iter` for every element.
+    #
+    # Internal details: the loop-carried primal `v` feeds no non-linear unary op so nothing requires
+    # `unary_collections` promotion, but the reverse pass still has to replay `v`'s per-iteration value via the
+    # adstack to compute the correct gradient. Breakage in the baseline adstack push/load/pop mechanics (missing
+    # push at the loop head, stale top-of-stack on LoadTop) would drive the result off by more than a ULP.
+    # `n_iter = 1` exercises the single-push adstack code path; `n_iter = 10` stresses repeated push/pop under one
+    # forward invocation; multi-element coverage (n = 4) guards against per-element accumulation bugs that a
+    # single-element variant would miss.
+    _run_basic_gradient(n_iter=n_iter, shall_not_pass=False)
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply tests/python/test_adstack.py]: Rewrote the `test_adstack_basic_gradient` "Internal details" block: it now says the adstack is structurally required for the dynamic loop (companion negative test pins that disabling adstack raises `QuadrantsCompilationError`), but value-correctness of the stored v is NOT exercised here because the linear backward chain `adj(v_prev) = 0.95 * adj(v_next)` only uses the constant 0.95 and never reads v from the adstack. Points readers at `test_adstack_unary_loop_carried` for push/load/pop value-correctness coverage.

File: tests/python/test_adstack.py [hunk 8]
```diff
@@ -143,3 +143,123 @@ def compute():
...
     assert math.isnan(x.grad[None])
+
+
+def _run_basic_gradient(n_iter, shall_not_pass):
+    # Builds the kernel, runs forward + backward, and asserts either a correct gradient (`shall_not_pass=False`,
+    # adstack enabled) or a compile-time rejection (`shall_not_pass=True`, adstack disabled). Split out so the
+    # positive test and negative test share one implementation. Loop body `v = v * 0.95 + 0.01` is loop-carried;
+    # reverse-mode AD needs the adstack to replay per-iteration `v` values - without the adstack the compiler
+    # refuses to build the backward kernel (`Cannot use non static range in Backwards mode`), which is exactly
+    # what the negative test pins.
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply tests/python/test_adstack.py]: Reworded the `_run_basic_gradient` helper comment. It now says the adstack is structurally required here so the backward compiler can reverse the dynamic `range(n_iter)` at all (without it the compiler refuses to build the backward kernel), and explicitly points at `test_adstack_basic_gradient`'s docstring for the note that value-correctness of per-iteration `v` is NOT exercised by the linear body. The two comments now agree.

**Comment by hughperkins** [comment]
Opus description update:

  ## Summary
  Adds three regression tests pinning baseline reverse-mode AD behavior through dynamic for-loops with the adstack extension. Test-only PR — no production code changes.
  The tests cover (a) that AD through a dynamic for-loop works at all and produces the analytical gradient, (b) that disabling the adstack causes the backward compiler to reject
   the same kernel shape at compile time rather than silently producing wrong gradients, and (c) that enabling the adstack extension does not silently regress linear
  reverse-mode AD across the static/dynamic × constant/varying-coefficient matrix.
  ## What's in the PR
  Net diff vs `origin/duburcqa/split_adjoint_alloca_placement`: **1 file, +121 / -0**.
  ### `_run_basic_gradient(n_iter, shall_not_pass)`
  Shared helper. Builds the kernel `for i: v = x[i]; for _ in range(n_iter): v = v*0.95 + 0.01; y += v`, runs forward + backward, and either asserts the analytical gradient
  `0.95 ** n_iter` (positive case) or asserts that `compute.grad()` raises `qd.QuadrantsCompilationError` matching `non static range` (negative case).
  ### `test_adstack_basic_gradient(n_iter ∈ {1, 3, 10})`
  Smallest "does reverse-mode AD through a for-loop work at all" check. Multi-element (`n = 4`) to catch per-element accumulation bugs. Docstring is explicit about scope: the
  adstack is *structurally* required to reverse the dynamic range, but value-correctness of the spilled `v` is *not* exercised because the loop body is linear — the backward
  chain `adj(v_prev) = 0.95 * adj(v_next)` only uses the compile-time constant 0.95 and never reads `v`. Cross-references `test_adstack_unary_loop_carried` for value-correctness
   coverage.
  ### `test_adstack_basic_gradient_negative(n_iter ∈ {1, 3, 10})`
  Negative counterpart, gated with `@test_utils.test(ad_stack_experimental_enabled=False)`. Pins that disabling the adstack causes the backward compiler to reject the same
  kernel with `QuadrantsCompilationError("Cannot use non static range in Backwards mode")`.
  ### `test_adstack_sum_linear(use_static_loop, use_varying_coeff, n_iter)`
  3 × 2 × 2 = 12 cases covering all combinations of (static-unrolled vs dynamic inner loop) × (constant vs loop-index-varying coefficient), at three trip counts. Replaces three
  earlier separate tests with one parametrized matrix. Uses the `qd.static(...) if qd.static(...) else ...` ternary idiom to switch loop kind. Docstring explains why no negative
   counterpart is included.
  ## Good points
  - **Pure regression coverage, no risk to production code.** The diff is one file of new tests; merge risk is essentially zero.
  - **Positive + negative pair.** `test_adstack_basic_gradient` and `test_adstack_basic_gradient_negative` lock in *both* the success path and the deterministic compile-time
  rejection on the same kernel shape. The negative test is what guards against the worst regression direction — adstack accidentally disabled producing silently-wrong gradients
  instead of a loud compile error.
  - **Honest scope documentation.** The `test_adstack_basic_gradient` docstring explicitly says what the test does *not* exercise (value-correctness of the spilled `v`) and why
  (the linear body's backward chain doesn't read `v`). It then points the reader at the test that *does* cover that gap. This is the kind of "what this test isn't" framing
  that's normally lost to PR comments and review threads; baking it into the source means a future contributor can't trivially over-rely on this test as a stronger oracle than
  it actually is.
  - **`test_adstack_sum_linear` is a parametrize matrix that subsumes three prior tests.** Net code is smaller than three hand-rolled variants would be, and every cell of the
  truth table is exercised at every trip count. Replaces a documentation problem (which test covers which combination?) with a parametrize problem (which is self-documenting).
  - **`n_iter ∈ {1, 3, 10}` consistently across all three tests.** `1` is the single-push edge case; `3`/`10` exercise repeated push/pop. Matches the convention from the
  baseline PR.
  - **Multi-element fields (`n = 4`).** Catches per-element accumulation bugs that single-element tests would silently miss.
  - **Match-string on the negative test (`r"non static range"`)** locks in the specific error condition, not just any `QuadrantsCompilationError`. Reduces the chance of a
  different compile error masquerading as the "expected" rejection.
  - **Iterative review history is visible in the commits.** Each follow-up commit (parametrize, multi-element, ternary idiom, error class refinement, docstring corrections)
  addresses a specific review point — easy to audit what changed and why.
  ## Bad points / concerns
  - **`test_adstack_sum_linear` is the weakest of the three.** Its docstring acknowledges this — "does not stress the adstack" — and declines to add a negative counterpart,
  which is defensible. But the value of "prove enabling the adstack extension does not silently regress linear AD" depends on linear AD not having its own per-loop-shape
  coverage elsewhere; if it does, this test is partly redundant. Worth a one-line cross-reference if such tests exist.
  - **No `f64` variant.** Same f32-only limitation as the rest of the chain. The linear-coefficient gradient `0.95 ** n_iter` is small enough at `n_iter = 10` (≈ 0.599) that f64
   wouldn't materially tighten the tolerance, but a smoke `f64` parametrize entry would still rule out a dtype-specific regression.
  - **Hardcoded `n = 4` and `x_vals = [0.1, 0.3, 0.5, 0.8]`** in both `_run_basic_gradient` and `test_adstack_sum_linear` — drifts if one is updated and the other isn't. A small
   shared constant would make the coupling explicit.
  - **`tol = 1e-4` (via `test_utils.approx(rel=1e-4)`)** matches the baseline PR's tolerance, but for a strictly linear kernel the gradient is exact in f32 up to round-off; a
  tighter tolerance (e.g. `rel=1e-6`) would be a stronger oracle and might catch a future bug that introduces extra fp ops in the backward.
  - **`test_adstack_basic_gradient_negative` only asserts the compile error fires.** It doesn't assert that *no* gradient gets computed (e.g. that `x.grad[i]` is unchanged from
  its pre-call zero). Probably fine because `pytest.raises` already implies the call didn't return, but a defensive post-assertion would harden the test against a future change
  where `compute.grad()` partially executes before raising.
  - **No coverage of nested dynamic loops.** All three tests use a single inner dynamic loop. The adstack's behavior under nested dynamic ranges is a meaningfully different code
   path. Probably out of scope here, but worth a follow-up.
  - **Inherits the merge-commit + multi-PR-stack process notes** from earlier PRs in the chain.

**Comment by hughperkins** [comment]
Please address:
- No f64 variant. Same f32-only limitation as the rest of the chain. The linear-coefficient gradient 0.95 ** n_iter is small enough at n_iter = 10 (≈ 0.599) that f64
wouldn't materially tighten the tolerance, but a smoke f64 parametrize entry would still rule out a dtype-specific regression.
- tol = 1e-4 (via test_utils.approx(rel=1e-4)) matches the baseline PR's tolerance, but for a strictly linear kernel the gradient is exact in f32 up to round-off; a
tighter tolerance (e.g. rel=1e-6) would be a stronger oracle and might catch a future bug that introduces extra fp ops in the backward.

**Comment by hughperkins** [comment]
Checklist:
- no code changes => doesn't change existing API or usage
    - no need for doc changes
- no code changes => doesn't add significantly more test code around diverse files

=> ok to merge
